### PR TITLE
Feature rtao

### PR DESCRIPTION
**Pull request by florianschut**
Added first version of ambient occlusion to hybrid composition

### CODE REVIEW

File: resources/shaders/deferred_composition.hlsl [hunk 1]
```diff
@@ -84,7 +85,8 @@ void main_cs(int3 dispatch_thread_id : SV_DispatchThreadID)
 			is_hybrid);
 
 		// Shade pixel
-		retval = shade_pixel(pos, V, albedo, metallic, roughness, normal, sampled_irradiance, reflection);
+		float3 occluded_irradiance = sampled_irradiance * gbuffer_AO[screen_coord].x;
+		retval = shade_pixel(pos, V, albedo, metallic, roughness, normal, occluded_irradiance, reflection);
```
Comment (MeineZ) [inline resources/shaders/deferred_composition.hlsl]: Use lerp instead (use sampled_arradiance) if it's not hybrid

File: src/resource_pool_texture.cpp [hunk 2]
```diff
@@ -22,6 +22,10 @@ namespace wr
 			{
 				texture = LoadPNG(path, srgb, generate_mips);
 			}
+			else if (ext_int.find("jpeg") != std::string_view::npos)
+			{
+				texture = LoadPNG(path, srgb, generate_mips);
+			}
```
Comment (MeineZ) [inline src/resource_pool_texture.cpp]: ext_int.find("jpeg") and then LoadPNG?

Comment (florianschut) [reply src/resource_pool_texture.cpp]: Yes, we use the png loader for jpeg

File: resources/shaders/deferred_composition.hlsl (hunk 1 above)
Comment (florianschut) [reply resources/shaders/deferred_composition.hlsl]: fixed!

**Comment by VZout** [comment]
Good work. Seems quite unstable though. Also I'd like a way to enable and disable the ambient occlusion.

**Comment by VZout** [comment]
it also breaks full raytracing.

File: src/resource_pool_texture.cpp [hunk 3]
```diff
@@ -22,6 +22,10 @@ namespace wr
 			{
 				texture = LoadPNG(path, srgb, generate_mips);
 			}
+			else if (ext_int.find("jpeg") != std::string_view::npos)
```
Comment (VZout) [inline src/resource_pool_texture.cpp]: find jpeg. loadpng. Something should be changed here :)

**Comment by florianschut** [comment]
> Good work. Seems quite unstable though. Also I'd like a way to enable and disable the ambient occlusion.

Implemented this for all RT features in the hybrid pipeline. Looking forward to some feedback on it

File: src/resource_pool_texture.cpp (hunk 3 above)
Comment (florianschut) [reply src/resource_pool_texture.cpp]: Fixed in [https://github.com/TeamWisp/Procedural-Ray-Tracing/pull/194/commits/3ff74f640cdf84cb53edc0bc8196ae274350507e](this commit)

**Comment by florianschut** [comment]
> See other comments and ambient occlusion should be part of a different render task. no need for it to be part of hybrid as far as I see.

I'll gladly look at it next sprint, this is not part of this sprint

**Comment by florianschut** [comment]
> it also breaks full raytracing.

Fixed, overall stability seems better aswell

File: src/d3d12/d3d12_resource_pool_texture.cpp [hunk 4]
```diff
@@ -402,7 +402,7 @@ namespace wr
 		delete texture;
 	}
 
-	d3d12::TextureResource* D3D12TexturePool::LoadPNG(std::string_view path, bool srgb, bool generate_mips)
+	d3d12::TextureResource* D3D12TexturePool::LoadPNG_JPEG(std::string_view path, bool srgb, bool generate_mips)
```
[the remark this answers is not in the document]
Comment (florianschut) [reply src/d3d12/d3d12_resource_pool_texture.cpp]: It's called LoadWIC from now on.

File: demo/demo_frame_graphs.hpp [hunk 5]
```diff
@@ -219,6 +234,17 @@ namespace fg_manager
 		}
 	}
 
+	inline void UpdateIfTasksChanged(wr::RenderSystem& rs, util::Delegate<void(ImTextureID)> imgui_func)
```
Comment (VZout) [inline demo/demo_frame_graphs.hpp]: Remove this please. I don't want this code in master since we can't test it correctly before the play day. I am not convinced rebuilding the SG works properly.

File: resources/shaders/ambient_occlusion.hlsl [hunk 6]
```diff
@@ -0,0 +1,12 @@
+#include "rt_global.hlsl"
+#include "util.hlsl"
+
+float DoAmbientOcclusion(float3 normal, float3 wpos, uint spp, float ray_length, uint rand_seed)
+{
+    float aoValue = 1.0f;
```
[the remark this answers is not in the document]
Comment (VZout) [reply resources/shaders/ambient_occlusion.hlsl]: still style guide...

File: resources/shaders/lighting.hlsl [hunk 7]
```diff
@@ -121,7 +121,13 @@ float3 shade_pixel(float3 pos, float3 V, float3 albedo, float metallic, float ro
 	[unroll]
 	for (uint i = 0; i < light_count; i++)
 	{
-		res += shade_light(pos, V, albedo, normal, metallic, roughness, lights[i], rand_seed, depth);
+		if(shadows == 1)
```
Comment (VZout) [inline resources/shaders/lighting.hlsl]: Why this change seems to be irrelevant to ao.

File: src/d3d12/d3d12_defines.hpp [hunk 8]
```diff
@@ -4,11 +4,29 @@
...
+
+	if (FormatMessage(
+		FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM,
+		NULL, hr, MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+		(LPTSTR)&szErrMsg, 0, NULL) != 0)
+	{
+		return(szErrMsg);
+		LocalFree(szErrMsg);
+	}
+	else
```
Comment (VZout) [inline src/d3d12/d3d12_defines.hpp]: braces. style guide

File: src/d3d12/d3d12_defines.hpp [hunk 9]
```diff
@@ -4,11 +4,29 @@
 
 #define D3DX12_INC d3dx12_rt.h
 
+inline std::string HResultToString(HRESULT hr)
+{
+	if (FACILITY_WINDOWS == HRESULT_FACILITY(hr))
```
Comment (VZout) [inline src/d3d12/d3d12_defines.hpp]: braces

File: src/d3d12/d3d12_renderer.hpp [hunk 10]
```diff
@@ -61,9 +63,11 @@ namespace wr
 			DirectX::XMMATRIX m_inverse_projection;
 			DirectX::XMMATRIX m_inv_vp;
 
-			uint32_t m_padding[2];
+			//uint32_t m_padding[0];
```
Comment (VZout) [inline src/d3d12/d3d12_renderer.hpp]: Don't comment things out. We are using version control.

File: src/imgui_tools.cpp [hunk 11]
```diff
@@ -19,8 +19,11 @@
 #include "d3d12/d3d12_pipeline_registry.hpp"
 #include "imgui/ImGuizmo.h"
 
+
 namespace wr::imgui::internal
 {
+	struct RTAOData;
```
Comment (VZout) [inline src/imgui_tools.cpp]: unused?

File: resources/shaders/lighting.hlsl (hunk 7 above)
Comment (florianschut) [reply resources/shaders/lighting.hlsl]: During development I was asked to add a toggle for AO, while implementing that I also implemented the same thing for shadows and reflections as it would save work in the end.

File: src/imgui_tools.cpp (hunk 11 above)
Comment (florianschut) [reply src/imgui_tools.cpp]: correct, was unused in the current implementation, to make things worse, what it was initially implemented is removed due to feedback on PR :feelsBadMan:

File: src/d3d12/d3d12_renderer.hpp (hunk 10 above)
Comment (florianschut) [reply src/d3d12/d3d12_renderer.hpp]: removed

**Comment by florianschut** [comment]
@VZout Thanks for the feedback, all of your points have been acted upon in https://github.com/TeamWisp/Procedural-Ray-Tracing/pull/194/commits/3f1eb84a92ee0df3e72251c6f35e7272f5016337

File: src/renderer.hpp [hunk 12]
```diff
@@ -68,6 +68,13 @@ namespace wr
 		virtual void Resize(std::uint32_t width, std::uint32_t height) = 0;
 		
 		std::optional<Window*> m_window;
+		
+		struct EffectsEnabled
```
Comment (VZout) [inline src/renderer.hpp]: The render system shouldn't be in charge of this. The user should be.

Comment (florianschut) [reply src/renderer.hpp]: Deprecated now that FG rebuilding is scrapped. Should not have been in the code base anymore, my bad! Should be fixed

File: src/scene_graph/camera_node.hpp [hunk 13]
```diff
@@ -42,6 +42,10 @@ namespace wr
 		DirectX::XMMATRIX m_inverse_projection;
 		DirectX::XMMATRIX m_inverse_view;
 
+		bool m_shadows_enabled = true;
+		bool m_reflections_enabled = true;
+		bool m_ao_enabled = true;
```
[the remark this answers is not in the document]
Comment (MeineZ) [reply src/scene_graph/camera_node.hpp]: Can this be removed yett?

**Comment by florianschut** [comment]
Currect state of the branch is a shitshow, do not merge!!!

File: src/scene_graph/camera_node.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (VZout) [reply src/scene_graph/camera_node.hpp]: Shouldn't be part of the camera though. The camera should be completely unrelated to render passes.

File: src/render_tasks/d3d12_deferred_composition.hpp [hunk 14]
```diff
@@ -26,11 +26,12 @@ namespace wr
 
 		bool is_hybrid;
 		bool is_path_tracer;
+		bool ao_enabled;
 	};
 
 	namespace internal
 	{
-
+		const unsigned int output_index = 7;
```
Comment (VZout) [inline src/render_tasks/d3d12_deferred_composition.hpp]: Hardcoded value? We have a system to help with this I believe.

File: src/scene_graph/camera_node.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (florianschut) [reply src/scene_graph/camera_node.hpp]: Negative, This is still needed so the composition shader doesn't sample from an empty buffer causing the entire scene to be occluded

File: src/render_tasks/d3d12_deferred_composition.hpp (hunk 14 above)
Comment (florianschut) [reply src/render_tasks/d3d12_deferred_composition.hpp]: Unused variable I accept all responsibilities but am clueless to why that's in there...

Comment (florianschut) [reply src/render_tasks/d3d12_deferred_composition.hpp]: Removed now

File: src/scene_graph/camera_node.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (florianschut) [reply src/scene_graph/camera_node.hpp]: I agree that this shouldn't be in the camera it simply was the easiest temporary way which I planned to overhaul next block anyway... Guess that refactor is moved forward, working on it right now.

File: demo/demo.cpp [hunk 15]
```diff
@@ -162,7 +162,9 @@ int WispEntry()
 
 	render_system->InitSceneGraph(*scene_graph.get());
 
-	fg_manager::Setup(*render_system, &RenderEditor);
+	bool is_fallback = wr::d3d12::GetRaytracingType(render_system->m_device) == wr::RaytracingType::FALLBACK;
```
Comment (VZout) [inline demo/demo.cpp]: The render system is already passed to the setup function, no need to get the is_fallback variable here.

File: resources/shaders/lighting.hlsl [hunk 16]
```diff
@@ -109,7 +109,7 @@ float3 shade_light(float3 pos, float3 V, float3 albedo, float3 normal, float met
 	return lighting;
 }
 
-float3 shade_pixel(float3 pos, float3 V, float3 albedo, float metallic, float roughness, float3 normal, inout uint rand_seed, uint depth)
+float3 shade_pixel(float3 pos, float3 V, float3 albedo, float metallic, float roughness, float3 normal, inout uint rand_seed, uint depth, uint shadows)
```
Comment (VZout) [inline resources/shaders/lighting.hlsl]: Why is shadows a uint?

File: src/d3d12/d3d12_defines.hpp [hunk 17]
```diff
@@ -6,11 +6,36 @@
...
+	}
+	TCHAR* sz_err_msg;
+
+	if (FormatMessage(
+		FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM,
+		NULL, hr, MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+		(LPTSTR)&sz_err_msg, 0, NULL) != 0)
+	{
+		return(sz_err_msg);
+		LocalFree(sz_err_msg);
```
Comment (VZout) [inline src/d3d12/d3d12_defines.hpp]: This code will never get called.

File: src/d3d12/d3d12_defines.hpp [hunk 18]
```diff
@@ -6,11 +6,36 @@
...
+
 //! Checks whether the d3d12 object exists before releasing it.
 #define SAFE_RELEASE(obj) { if ( obj ) { obj->Release(); obj = NULL; } }
 
 //! Handles a hresult.
-#define TRY(result) if (FAILED(result)) { LOGC("An hresult returned a error!. File: " + std::string(__FILE__) + " Line: " + std::to_string(__LINE__)); }
+#define TRY(result) if (FAILED(result)) { LOGC("An hresult returned a error!. File: " + std::string(__FILE__) + " Line: " + std::to_string(__LINE__) + " HRResult: " +  HResultToString(result)); }
 
 //! Handles a hresult and outputs a specific message.
 #define TRY_M(result, msg) if (FAILED(result)) { LOGC(msg); }
```
Comment (VZout) [inline src/d3d12/d3d12_defines.hpp]: TRY_M could also make use of the `HResultToString` function.

File: resources/shaders/ambient_occlusion.hlsl [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+	float4x4 inv_view;
+	float4x4 inv_projection;
+	float4x4 inv_vp;
+
+	float2 padding;
+	float frame_idx;
+	float intensity;
+};
+
+struct Attributes { };
```
Comment (LaisoEmilio) [inline resources/shaders/ambient_occlusion.hlsl]: ????

File: src/render_tasks/d3d12_rtao_task.hpp [hunk 20]
```diff
@@ -0,0 +1,292 @@
...
+			auto& data = fg.GetData<RTAOData>(handle);
+			auto n_render_target = fg.GetRenderTarget<d3d12::RenderTarget>(handle);
+			d3d12::SetName(n_render_target, L"AO Target");
+
+			if (!resize)
+			{
+				auto cmd_list = fg.GetCommandList<d3d12::CommandList>(handle);
+				auto pred_cmd_list = fg.GetPredecessorCommandList<wr::ASBuildData>();
+
+				cmd_list->m_rt_descriptor_heap = static_cast<d3d12::CommandList*>(pred_cmd_list)->m_rt_descriptor_heap;
```
Comment (LaisoEmilio) [inline src/render_tasks/d3d12_rtao_task.hpp]: this is very old code, we changed the logic. Look at the current master version of the hybrid render task. Each command list now uses its own desc heap, we don't have to do this anymore. Also, due to this fact, the execution task in hybrid now builds the acceleration structure on the descriptor heap, therefore the AO task should do the same.

File: resources/shaders/deferred_composition.hlsl [hunk 21]
```diff
@@ -16,8 +16,8 @@ TextureCube irradiance_map   : register(t5);
 TextureCube pref_env_map	 : register(t6);
 Texture2D brdf_lut			 : register(t7);
 Texture2D buffer_refl_shadow : register(t8); // xyz: reflection, a: shadow factor
-Texture2D screen_space_irradiance : register(t9);
-Texture2D screen_space_ao : register(t10);
+Texture2D gbuffer_AO 		 : register(t9); //x: AO value
```
Comment (VZout) [inline resources/shaders/deferred_composition.hlsl]: This is not a gbuffer :O. the old name is good enough.

Comment (florianschut) [reply resources/shaders/deferred_composition.hlsl]: I adopted the old name, misunderstood the use of screen_space in the naming, also had not clue what a GBuffer actually was 😨

File: src/render_tasks/d3d12_rtao_task.hpp (hunk 20 above)
Comment (florianschut) [reply src/render_tasks/d3d12_rtao_task.hpp]: Sould be updated

File: src/render_tasks/d3d12_deferred_composition.cpp [hunk 22]
```diff
@@ -143,8 +145,17 @@ namespace wr
 
 					auto hybrid_rt = static_cast<d3d12::RenderTarget*>(fg.GetPredecessorRenderTarget<wr::RTHybridData>());
 					d3d12::CreateSRVFromRTV(hybrid_rt, shadow_handle, 1, hybrid_rt->m_create_info.m_rtv_formats.data());
-				}
-			}
+				
+					if (data.is_rtao)
+					{
+						constexpr auto ao_id = rs_layout::GetHeapLoc(params::deferred_composition, params::DeferredCompositionE::BUFFER_AO);
```
Comment (EmilioLaiso) [inline src/render_tasks/d3d12_deferred_composition.cpp]: Not used

File: src/render_tasks/d3d12_deferred_composition.cpp [hunk 23]
```diff
@@ -143,8 +145,17 @@ namespace wr
 
 					auto hybrid_rt = static_cast<d3d12::RenderTarget*>(fg.GetPredecessorRenderTarget<wr::RTHybridData>());
```
Comment (EmilioLaiso) [inline src/render_tasks/d3d12_deferred_composition.cpp]: Not used, even if not yours

File: src/render_tasks/d3d12_rtao_task.hpp [hunk 24]
```diff
@@ -0,0 +1,297 @@
...
+			d3d12::CreateOrUpdateTLAS(device, cmd_list, data.tlas_requires_init, data.out_tlas, as_build_data.out_blas_list);
+
+			// Wait for AS to be built
+			{
+				auto barrier = CD3DX12_RESOURCE_BARRIER::UAV(as_build_data.out_tlas.m_native);
+				cmd_list->m_native->ResourceBarrier(1, &barrier);
+			}
+
+			// Wait for AS to be built
+			cmd_list->m_native->ResourceBarrier(1, &CD3DX12_RESOURCE_BARRIER::UAV(as_build_data.out_tlas.m_native));
```
Comment (EmilioLaiso) [inline src/render_tasks/d3d12_rtao_task.hpp]: double resource barrier

Comment (florianschut) [reply src/render_tasks/d3d12_rtao_task.hpp]: This is indeed duplicate, though when looking at it I found something interesting. This code was inspired by the hybrid render task which is as follows: 
```cpp
// Rebuild acceleration structure a 2e time for fallback
			if (d3d12::GetRaytracingType(device) == RaytracingType::FALLBACK)
			{
				d3d12::CreateOrUpdateTLAS(device, cmd_list, data.tlas_requires_init, data.out_tlas, as_build_data.out_blas_list);

				auto barrier = CD3DX12_RESOURCE_BARRIER::UAV(as_build_data.out_tlas.m_native);
				cmd_list->m_native->ResourceBarrier(1, &barrier);
			}```
Yet fallback is not supported at all in RTAO, therefore I'd like to remove the resource barier entirely and move the fallback check from the demo framegraph to inside the AddAOTask (which should be renamed to AddRTAOTask while i think about it).

@VZout Will this implementation I described be correct or am I missing something?
